libs/galley-types/src/Galley/Types/Teams.hs

```diff
@@ -149,13 +149,26 @@ rolePermissions :: Role -> Permissions
 rolePermissions role = Permissions p p where p = rolePerms role
 
 permissionsRole :: Permissions -> Maybe Role
-permissionsRole (Permissions p p') | p /= p' = Nothing
-permissionsRole (Permissions p _) = permsRole p
+permissionsRole (Permissions p p') =
+  if p /= p'
+    then do
+      -- we never did use @p /= p'@ for anything, fingers crossed that it doesn't occur anywhere
+      -- in the wild.  but if it does, this implementation prevents privilege escalation.
+      let p'' = Set.intersection p p'
+       in permissionsRole (Permissions p'' p'')
+    else permsRole p
   where
     permsRole :: Set Perm -> Maybe Role
     permsRole perms =
       Maybe.listToMaybe
-        [role | role <- [minBound ..], rolePerms role == perms]
+        [ role
+          | role <- [minBound ..],
```

Comment:
```suggestion
          | role <-  reverse [minBound ..],
```

Rereading this this function is now incorrect I think.  It now gives the least permissive role that is a subset of `perms` but we want the most permissive role that is still a subset.   The current implementation maps all users to a `rolePerms minBound` as  `rolePerms minBound` is trivially the subset of almost all permission sets in the wild.

Reply:
the tests would have caught it, but you're wrong about `Role`:

```
*Galley.Types.Teams> [minBound..]::[Role]
[RoleOwner,RoleAdmin,RoleMember,RoleExternalPartner]
```

Reply:
Ah so it is ordered exactly opposite of what I expected. All good then :)